src/main/java/com/google/devtools/build/lib/analysis/starlark/StarlarkRuleContext.java

```diff
@@ -701,9 +702,24 @@ public StarlarkAttributesCollection rule() throws EvalException {
   }
 
   @Override
-  public ResolvedToolchainContext toolchains() throws EvalException {
+  public ToolchainContextApi toolchains() throws EvalException {
     checkMutable("toolchains");
-    return ruleContext.getToolchainContext();
+    ResolvedToolchainContext toolchainContext = ruleContext.getToolchainContext();
+    if (toolchainContext == null) {
+      // Starlark rules are easier if this cannot be null, so return a no-op value instead.
+      return new ToolchainContextApi() {
```

Comment:
Optionally make this a singleton, but I doubt it'll matter much.

Reply:
I don't think there's enough benefit, this is fairly clear.